The `copy_data_from` feature doesn't work on MySQL 8

I know that. Even if we delete the `--ignore-data-dir` option, the following error will occur.

```
2018-07-19T13:39:48.694776Z 0 [ERROR] [MY-010457] [Server] --initialize specified but the data directory has files in it. Aborting.
```

--- a/META.json
+++ b/META.json
@@ -33,6 +33,7 @@
       },
       "develop" : {
          "requires" : {
+            "Pod::Markdown::Github" : "0",
             "Test::CPAN::Meta" : "0",
             "Test::MinimumVersion::Fast" : "0.04",
             "Test::PAUSE::Permissions" : "0.04",

--- a/lib/Test/mysqld.pm
+++ b/lib/Test/mysqld.pm
@@ -145,7 +145,6 @@ sub wait_for_setup {
         $dbh->do('CREATE DATABASE IF NOT EXISTS test')
             or die $dbh->errstr;
     }
-    # XXX copy_data_from doesn't work on MySQL8
 }
 
 sub stop {
@@ -189,6 +188,12 @@ sub setup {
         dircopy($self->copy_data_from, $self->my_cnf->{datadir})
             or die "could not dircopy @{[$self->copy_data_from]} to " .
                 "@{[$self->my_cnf->{datadir}]}:$!";
+        if (!$self->_is_maria && ($self->_mysql_major_version || 0) >= 8) {
+            my $mysql_db_dir = $self->my_cnf->{datadir} . '/mysql';
+            if (! -d $mysql_db_dir) {
+                mkdir $mysql_db_dir or die "failed to mkdir $mysql_db_dir: $!";
+            }
+        }
     }
     # my.cnf
     open my $fh, '>', $self->base_dir . '/etc/my.cnf'
@@ -217,7 +222,9 @@ sub setup {
 
         if ($self->use_mysqld_initialize) {
             $cmd .= ' --initialize-insecure';
-            if ($self->copy_data_from) {
+            if ($self->copy_data_from &&
+                !(!$self->_is_maria && ($self->_mysql_major_version || 0) >= 8)
+            ) {
                 opendir my $dh, $self->copy_data_from
                     or die "failed to open copy_data_from directory @{[$self->copy_data_from]}: $!";
                 while (my $entry = readdir $dh) {

--- a/minil.toml
+++ b/minil.toml
@@ -4,3 +4,6 @@ authority="cpan:KAZUHO"
 
 module_maker="ModuleBuildTiny"
 markdown_maker = "Pod::Markdown::Github"
+
+[FileGatherer]
+exclude_match = ['^xt/copy-data-from-mysql8*', '^author/*', 'minil.toml']

--- a/t/05-copy-data-from.t
+++ b/t/05-copy-data-from.t
@@ -4,35 +4,67 @@ use warnings;
 use DBI;
 use Test::More;
 use Test::mysqld;
+use File::Temp qw/tempdir/;
+use File::Copy qw/copy/;
 
 my $mysqld = Test::mysqld->new(
     auto_start     => undef,
-    copy_data_from => 't/05-copy-data-from',
     my_cnf         => {
         'skip-networking' => '',
     },
 ) or plan skip_all => $Test::mysqld::errstr;
 
-plan tests => 1;
-
-TODO: {
-    my $ver = $mysqld->_mysql_major_version;
-    todo_skip "The `copy_data_from` feature doesn't work on MySQL 8", 1
-        if !$mysqld->_is_maria && $ver && $ver >= 8;
-
+if (!$mysqld->_is_maria && ($mysqld->_mysql_major_version || 0) >= 8) {
     $mysqld->setup;
     $mysqld->start;
-
     my $dbh = DBI->connect($mysqld->dsn)
         or die "failed to connect to database:$DBI::errstr";
 
-    is_deeply(
-        $dbh->selectall_arrayref(
-            "select id,str from test.hello order by id",
-        ),
-        [
-            [ 1, 'hello' ],
-            [ 2, 'ciao' ],
-        ],
-    );
+    $dbh->do('CREATE TABLE hello (
+      id INTEGER PRIMARY KEY,
+      str VARCHAR(191)
+    )') or die $DBI::errstr;
+
+    $dbh->do(q{INSERT INTO hello VALUES (
+      1, 'hello'
+    ), (
+      2, 'ciao'
+    )}) or die $DBI::errstr;
+    $mysqld->stop;
+
+    my $copydir = tempdir(CLEANUP => 1);
+    mkdir $copydir . '/test';
+    for my $d (qw(/ibdata1 /mysql.ibd /test/hello.ibd)) {
+        copy $mysqld->my_cnf->{datadir} . $d, $copydir . $d;
+    }
+    undef $mysqld;
+
+    $mysqld = Test::mysqld->new(
+        copy_data_from => $copydir,
+        auto_start     => undef,
+        my_cnf         => {
+            'skip-networking' => '',
+        },
+    ) or plan skip_all => $Test::mysqld::errstr;
 }
+else {
+    $mysqld->copy_data_from('t/05-copy-data-from');
+}
+
+plan tests => 1;
+
+$mysqld->setup;
+$mysqld->start;
+
+my $dbh = DBI->connect($mysqld->dsn)
+    or die "failed to connect to database:$DBI::errstr";
+
+is_deeply(
+    $dbh->selectall_arrayref(
+        "select id,str from test.hello order by id",
+    ),
+    [
+        [ 1, 'hello' ],
+        [ 2, 'ciao' ],
+    ],
+);

